### ISSUE TITLE
Person reports: Ability to filter date fields before AND after the current date

### ISSUE THREAD

**Issue by tbar0970** [tbar0970/jethro-pmm#163]
Currently you can filter by a date field value being within X days either before OR after the date on which the report is run.
It would be valuable to be able to filter by, for example, from 30 days before til 30 days after the report date.

**Comment by tbar0970** [comment]
Also to filter for dates that are MORE THAN x days before/after the current date.
Eg training dates more than 365 days ago.

**Comment by vanoudt** [comment]
I have a local branch where I am currently replacing all of the date fields in mysql to be DATE fields rather than text fields. Things like this are then very very easy to do using plain SQL.... looking at the SQL produced by reports, using DATE types for dates drastically reduces the complexity of the SQL, which can only make the queries much faster.

For dates that don't have a year associated with them, I simply set a very early leap year as the year, and in php filter that out/in as required...

**Comment by tbar0970** [comment]
Hmm, interesting.  Your queries would still need to have special handling for your "magic year" dates, eg if I'm looking for exact value or anniversary in the last 30 days, I'd need to figure out the relevant period within the "magic year" and filter for that too.  Which gets complex if it's 10th January and I'm searching for dates or anniversaries in the last 30 days.  

Which is to say:  I'm not sure that the existing complexity is caused by the fact that the column isn't of type DATE - I think it's inherent to the situation of having year-less dates in the mix.  MySQL is quite happy to cast values as dates along the way anyway.

I don't think this particular FR would increase the complexity of the queries - it just requires that the interface be extended to allow for more complex offsets.  Making that possible whilst keeping the interface easy to understand will be the hard bit.

But maybe I'm missing something...

**Comment by vanoudt** [comment]
Actually, one can use the MYSQL DAYOFYEAR function to easily do things like
that. The "magic year" is simply one that is a leap year, so DAYOFYEAR
stuff should work ok....

Agreed that the current TEXT type works, and casting is definitely useful,
but being able to use the MYSQL functions (like DAYOFYEAR) makes things a
whole bunch easier... In my report test, the SQL WHERE clause drops to
about a third of the size - easier reading, easier to understand etc. etc.




------------------------------

**Comment by tbar0970** [comment]
I'll be interested to have a look :-)

**Comment by vanoudt** [comment]
Just threw up the current code as a pull request...

Have a look at the person_query.class.php, around line 920. We go from:
`							$qFromYear = $db->quote(substr($from, 0, 4));
-								$qToYear = $db->quote(substr($to, 0, 4));
-
-								$w[] = "$valExp LIKE '-%' AND (
-											CONCAT($qFromYear, $valExp) $betweenExp
-											OR CONCAT($qToYear, $valExp) $betweenExp
-										)";
-								$w[] = "$valExp NOT LIKE '-%' AND (
-											CONCAT($qFromYear, RIGHT($valExp, 6)) $betweenExp
-											OR CONCAT($qToYear, RIGHT($valExp, 6)) $betweenExp
-										)";`
to
` +								$w[] = "DAYOFYEAR($valExp) BETWEEN DAYOFYEAR(".$db->quote($from) . ") AND DAYOFYEAR(".$db->quote($to).")"; `

**Comment by vanoudt** [comment]
And don't have to have the OR for years without dates anymore...

**Comment by vanoudt** [comment]
Just to make it clear: that pull request is by _no_ means actually ready for pulling - just there for demonstrating the approach!

**Comment by tbar0970** [comment]
Yes, using a magic-value year simplifies the query, but I still don’t think DAYOFYEAR is going to work across year boundaries:

If today is 10th Jan and I’m interested in the last 30 days, then in approximate numbers we get

DAYOFYEAR($valExp) BETWEEN DAYOFYEAR(‘2016-12-10’) AND DAYOFYEAR(‘2017-01-10’)
which equates to
DAYOFYEAR($valExp) BETWEEN 345 AND 10
which will always be false - https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_between <https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_between>

**Comment by vanoudt** [comment]
Hmmmm - that's a very good point.... hadn't thought of that....

there would have to be special consideration given to that sort of case..


------------------------------
